File: python/tests/test_base64plus.py
"""
Unit tests for the Base64Plus Python package.
"""

import os
import json
import unittest
from io import BytesIO
from PIL import Image

# Try to import the package
try:
    from base64plus import encode_base64plus, decode_base64plus
    PACKAGE_INSTALLED = True
except ImportError:
    # If package is not installed, try to import from local path
    import sys
    import os.path
    sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
    try:
        from base64plus import encode_base64plus, decode_base64plus
        PACKAGE_INSTALLED = True
    except ImportError:
        PACKAGE_INSTALLED = False


class TestBase64Plus(unittest.TestCase):
    """Test cases for Base64Plus functionality."""
    
    @classmethod
    def setUpClass(cls):
        """Set up test fixtures."""
        # Create a simple test image
        cls.test_image_path = os.path.join(os.path.dirname(__file__), 'test_image.png')
        
        # Create a simple image with text
        img = Image.new('RGB', (300, 100), color=(255, 255, 255))
        from PIL import ImageDraw, ImageFont
        draw = ImageDraw.Draw(img)
        
        # Try to use a font that's likely to be available
        try:
            font_paths = [
                '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf',  # Linux
                '/Library/Fonts/Arial.ttf',  # macOS
                'C:\\Windows\\Fonts\\arial.ttf',  # Windows
                '/System/Library/Fonts/Helvetica.ttc'  # macOS alternative
            ]
            
            font = None
            for font_path in font_paths:
                if os.path.exists(font_path):
                    font = ImageFont.truetype(font_path, 24)
                    break
            
            if font is None:
                font = ImageFont.load_default()
        except Exception:
            font = ImageFont.load_default()
        
        # Draw text
        draw.text((10, 10), "Base64Plus Test", fill=(0, 0, 0), font=font)
        draw.text((10, 50), "Testing 123", fill=(0, 0, 0), font=font)
        
        # Save the image
        img.save(cls.test_image_path)
    
    @classmethod
    def tearDownClass(cls):
        """Clean up test fixtures."""
        # Remove the test image
        if os.path.exists(cls.test_image_path):
            os.remove(cls.test_image_path)
    
    def test_package_installed(self):
        """Test that the package is installed."""
        self.assertTrue(PACKAGE_INSTALLED, "Base64Plus package is not installed")
    
    def test_encode_decode_roundtrip(self):
        """Test encoding and decoding an image."""
        if not PACKAGE_INSTALLED:
            self.skipTest("Base64Plus package is not installed")
        
        # Skip if OCR engines are not available
        try:
            # Encode the test image
            base64plus_data = encode_base64plus(self.test_image_path)
            
            # Check that the result is valid JSON
            data = json.loads(base64plus_data)
            
            # Check that the required fields are present
            self.assertIn('image', data, "Encoded data missing 'image' field")
            self.assertIn('text_data', data, "Encoded data missing 'text_data' field")
            self.assertIn('format', data, "Encoded data missing 'format' field")
            
            # Decode the data
            image, text_data = decode_base64plus(base64plus_data)
            
            # Check that the image is a PIL Image
            self.assertIsInstance(image, Image.Image, "Decoded image is not a PIL Image")
            
            # Check that text_data is a list
            self.assertIsInstance(text_data, list, "Decoded text_data is not a list")
            
            # If OCR worked, there should be some text data
            # But we can't guarantee this, so we don't assert on the content
            
        except Exception as e:
            if "No OCR engine available" in str(e):
                self.skipTest("No OCR engine available")
            else:
                raise
    
    def test_encode_with_options(self):
        """Test encoding with various options."""
        if not PACKAGE_INSTALLED:
            self.skipTest("Base64Plus package is not installed")
        
        # Skip if OCR engines are not available
        try:
            # Test with different options
            options_to_test = [
                {'include_confidence': True, 'image_format': 'PNG'},
                {'include_confidence': False, 'image_format': 'JPEG'},
            ]
            
            for options in options_to_test:
                # Encode with options
                base64plus_data = encode_base64plus(
                    self.test_image_path,
                    include_confidence=options['include_confidence'],
                    image_format=options['image_format']
                )
                
                # Check that the result is valid JSON
                data = json.loads(base64plus_data)
                
                # Check format
                self.assertEqual(
                    data['format'].lower(), 
                    options['image_format'].lower(),
                    f"Incorrect format: expected {options['image_format'].lower()}, got {data['format'].lower()}"
                )
                
                # Check confidence
                if not options['include_confidence'] and len(data['text_data']) > 0:
                    # If confidence should be excluded and we have text data,
                    # check that no item has a confidence field
                    for item in data['text_data']:
                        self.assertNotIn(
                            'confidence', item,
                            "Confidence field present when include_confidence=False"
                        )
        
        except Exception as e:
            if "No OCR engine available" in str(e):
                self.skipTest("No OCR engine available")
            else:
                raise


if __name__ == '__main__':
    unittest.main()

File: javascript/tests/test-base64plus.js
/**
 * Unit tests for the Base64Plus JavaScript package.
 * 
 * To run these tests:
 * 1. Make sure you have installed the package dependencies
 * 2. Run: node tests/test-base64plus.js
 */

const fs = require('fs');
const path = require('path');
const assert = require('assert');
const { createCanvas, loadImage } = require('canvas');

// Import the Base64Plus library
let base64plus;
try {
  base64plus = require('../index');
  console.log('✓ Successfully imported Base64Plus library');
} catch (error) {
  console.error('✗ Failed to import Base64Plus library:', error.message);
  process.exit(1);
}

// Test image path
const TEST_IMAGE_PATH = path.join(__dirname, 'test_image.png');

/**
 * Create a simple test image with text
 */
async function createTestImage() {
  // Create a canvas
  const canvas = createCanvas(300, 100);
  const ctx = canvas.getContext('2d');
  
  // Fill with white background
  ctx.fillStyle = 'white';
  ctx.fillRect(0, 0, 300, 100);
  
  // Add some text
  ctx.fillStyle = 'black';
  ctx.font = '24px Arial';
  ctx.fillText('Base64Plus Test', 10, 30);
  ctx.fillText('Testing 123', 10, 70);
  
  // Save to file
  const out = fs.createWriteStream(TEST_IMAGE_PATH);
  const stream = canvas.createPNGStream();
  stream.pipe(out);
  
  return new Promise((resolve, reject) => {
    out.on('finish', resolve);
    out.on('error', reject);
  });
}

/**
 * Clean up test files
 */
function cleanup() {
  if (fs.existsSync(TEST_IMAGE_PATH)) {
    fs.unlinkSync(TEST_IMAGE_PATH);
    console.log('✓ Cleaned up test image');
  }
}

/**
 * Run all tests
 */
async function runTests() {
  console.log('Running Base64Plus JavaScript tests...\n');
  
  try {
    // Create test image
    console.log('Creating test image...');
    await createTestImage();
    console.log('✓ Test image created');
    
    // Test 1: Encode and decode roundtrip
    console.log('\nTest 1: Encode and decode roundtrip');
    try {
      // Encode the test image
      const base64PlusData = await base64plus.encodeBase64Plus(TEST_IMAGE_PATH);
      console.log('✓ Successfully encoded image');
      
      // Check that the result is valid JSON
      const data = JSON.parse(base64PlusData);
      console.log('✓ Encoded data is valid JSON');
      
      // Check that the required fields are present
      assert(data.image, "Encoded data missing 'image' field");
      assert(data.text_data, "Encoded data missing 'text_data' field");
      assert(data.format, "Encoded data missing 'format' field");
      console.log('✓ Encoded data has all required fields');
      
      // Decode the data
      const decoded = await base64plus.decodeBase64Plus(base64PlusData);
      console.log('✓ Successfully decoded data');
      
      // Check that the image buffer exists
      assert(Buffer.isBuffer(decoded.imageBuffer), "Decoded image is not a Buffer");
      console.log('✓ Decoded image is a valid Buffer');
      
      // Check that text_data is an array
      assert(Array.isArray(decoded.textData), "Decoded text_data is not an array");
      console.log('✓ Decoded text_data is an array');
      
      // If OCR worked, there should be some text data
      if (decoded.textData.length > 0) {
        console.log(`✓ OCR detected ${decoded.textData.length} text elements`);
        
        // Log the first text element
        const firstText = decoded.textData[0];
        console.log(`  First text: "${firstText.text}" at (${firstText.x}, ${firstText.y})`);
      } else {
        console.log('⚠ OCR did not detect any text (this might be normal depending on OCR quality)');
      }
      
      console.log('✓ Test 1 passed');
    } catch (error) {
      console.error('✗ Test 1 failed:', error.message);
      throw error;
    }
    
    // Test 2: Encode with options
    console.log('\nTest 2: Encode with options');
    try {
      // Test with different options
      const options = { includeConfidence: false, imageFormat: 'jpeg' };
      
      // Encode with options
      const base64PlusData = await base64plus.encodeBase64Plus(TEST_IMAGE_PATH, options);
      console.log('✓ Successfully encoded image with options');
      
      // Check that the result is valid JSON
      const data = JSON.parse(base64PlusData);
      console.log('✓ Encoded data is valid JSON');
      
      // Check format
      assert.equal(
        data.format.toLowerCase(),
        options.imageFormat.toLowerCase(),
        `Incorrect format: expected ${options.imageFormat.toLowerCase()}, got ${data.format.toLowerCase()}`
      );
      console.log(`✓ Image format is correct: ${data.format}`);
      
      // Check confidence
      if (!options.includeConfidence && data.text_data.length > 0) {
        // If confidence should be excluded and we have text data,
        // check that no item has a confidence field
        for (const item of data.text_data) {
          assert(!item.hasOwnProperty('confidence'), "Confidence field present when includeConfidence=false");
        }
        console.log('✓ Confidence field correctly excluded');
      }
      
      console.log('✓ Test 2 passed');
    } catch (error) {
      console.error('✗ Test 2 failed:', error.message);
      throw error;
    }
    
    // Test 3: Render with bounding boxes
    console.log('\nTest 3: Render with bounding boxes');
    try {
      // Encode the test image
      const base64PlusData = await base64plus.encodeBase64Plus(TEST_IMAGE_PATH);
      
      // Decode the data
      const decoded = await base64plus.decodeBase64Plus(base64PlusData);
      
      // Render with bounding boxes
      const outputPath = path.join(__dirname, 'visualization.png');
      await base64plus.renderWithBoundingBoxes(decoded, outputPath);
      console.log('✓ Successfully rendered image with bounding boxes');
      
      // Check that the file exists
      assert(fs.existsSync(outputPath), "Visualization file was not created");
      console.log('✓ Visualization file was created');
      
      // Clean up the visualization file
      fs.unlinkSync(outputPath);
      console.log('✓ Cleaned up visualization file');
      
      console.log('✓ Test 3 passed');
    } catch (error) {
      console.error('✗ Test 3 failed:', error.message);
      throw error;
    }
    
    console.log('\n✅ All tests passed!');
  } catch (error) {
    console.error('\n❌ Tests failed:', error);
    process.exitCode = 1;
  } finally {
    // Clean up
    cleanup();
  }
}

// Run the tests
runTests();